jinja2/nativetypes.py

```diff
@@ -0,0 +1,211 @@
+# -*- coding: utf-8 -*-
+"""
+    jinja2.nativetypes
+    ~~~~~~~~~~~~~~~~~~
+
+    Provides an Environment to preserve native python variable types.
+
+    :copyright: (c) 2017 by the Jinja Team.
+    :license: BSD, see LICENSE for more details.
+"""
+
+import sys
+from ast import literal_eval
+from itertools import islice, chain
+from jinja2 import nodes
+from jinja2._compat import text_type
+from jinja2.compiler import CodeGenerator
+from jinja2.environment import Environment, Template
+from jinja2.utils import concat, escape
+
+
+def native_concat(nodes):
+    """Jinja's native concat is just a list join. The list comes from result of
+    templating all the nodes. If the return type is expected to be
+    native, the list is artificial and we should return just the first
+    element.
+    """
+    head = list(islice(nodes, 2))
+
+    if not head:
+        return None
+
+    if len(head) == 1:
+        out = head[0]
+    else:
+        out = u''.join([text_type(v) for v in chain(head, nodes)])
+
+    try:
+        return literal_eval(out)
+    except (ValueError, SyntaxError, MemoryError):  # possibly RecursionError
+        return out
+
+
+def noop_finalize(node):
+    """Do nothing as a fallback when no finalize method is defined
+    """
+    return node
+
+
+class NativeCodeGenerator(CodeGenerator):
+    """A custom code generator, which avoids injecting to_string() calls around
+    the interal code jinja uses to render templates.
+    """
+
+    def visit_Output(self, node, frame):
+        """Slightly modified from the same method in :class:CodeGenerator,
```

Comment:
"My eventual goal is to see if the jinja dev team would consider allowing me to break up CodeGenerator.visitOutput into a few smaller functions. "